### PR TITLE
Update Dockerfile to ubuntu:focal (Python 3.8)

### PR DESCRIPTION
**Pull request by manics**
This updates the Docker base image from Ubuntu bionic (18.04) to focal (20.04), which bumps Python3 to 3.8. This might help deal with problems with SameSite cookies:
- https://github.com/jupyterhub/jupyterhub/issues/3117
- https://discourse.jupyter.org/t/issue-with-chrome-80-and-samesite-cookie-changes/3757/7

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -21,8 +21,7 @@
 # your jupyterhub_config.py will be added automatically
 # from your docker directory.
 
-# https://github.com/tianon/docker-brew-ubuntu-core/commit/d4313e13366d24a97bd178db4450f63e221803f1
```
Comment (manics) [inline Dockerfile:24]: I couldn't figure out how to get the corresponding commit for `ubuntu:focal-20200729@sha256:6f2fb2f9fb5582f8b587837afd6ea8f37d8d1d9e41168c90f410a6ef15fa8ce5`

**Comment by manics** [comment]
This isn't as straightforward as I thought it'd be 😥:
```
$ docker run -it --rm -p8000:8000 jupyterhub
[I 2020-08-26 21:34:50.334 JupyterHub app:2307] Running JupyterHub version 1.2.0dev
[I 2020-08-26 21:34:50.335 JupyterHub app:2337] Using Authenticator: jupyterhub.auth.PAMAuthenticator-1.2.0dev
[I 2020-08-26 21:34:50.335 JupyterHub app:2337] Using Spawner: jupyterhub.spawner.LocalProcessSpawner-1.2.0dev
[I 2020-08-26 21:34:50.335 JupyterHub app:2337] Using Proxy: jupyterhub.proxy.ConfigurableHTTPProxy-1.2.0dev
[I 2020-08-26 21:34:50.344 JupyterHub app:1442] Writing cookie_secret to /srv/jupyterhub/jupyterhub_cookie_secret
[I 2020-08-26 21:34:50.363 alembic.runtime.migration migration:155] Context impl SQLiteImpl.
[I 2020-08-26 21:34:50.363 alembic.runtime.migration migration:158] Will assume non-transactional DDL.
[I 2020-08-26 21:34:50.375 alembic.runtime.migration migration:515] Running stamp_revision  -> 4dc2d5a8c53c
[I 2020-08-26 21:34:50.430 JupyterHub proxy:461] Generating new CONFIGPROXY_AUTH_TOKEN
[W 2020-08-26 21:34:50.431 JupyterHub app:1672] No admin users, admin interface will be unavailable.
[W 2020-08-26 21:34:50.431 JupyterHub app:1673] Add any administrative users to `c.Authenticator.admin_users` in config.
[I 2020-08-26 21:34:50.431 JupyterHub app:1702] Not using allowed_users. Any authenticated user will be allowed.
[I 2020-08-26 21:34:50.471 JupyterHub app:2374] Initialized 0 spawners in 0.002 seconds
[W 2020-08-26 21:34:50.474 JupyterHub proxy:642] Running JupyterHub without SSL.  I hope there is SSL termination happening somewhere else...
[I 2020-08-26 21:34:50.474 JupyterHub proxy:646] Starting proxy @ http://:8000
21:34:51.077 [ConfigProxy] info: Proxying http://*:8000 to (no default)
21:34:51.083 [ConfigProxy] info: Proxy API at http://127.0.0.1:8001/api/routes
21:34:51.577 [ConfigProxy] info: 200 GET /api/routes
[I 2020-08-26 21:34:51.579 JupyterHub app:2622] Hub API listening on http://127.0.0.1:8081/hub/
21:34:51.581 [ConfigProxy] info: 200 GET /api/routes
[I 2020-08-26 21:34:51.581 JupyterHub proxy:320] Checking routes
[I 2020-08-26 21:34:51.581 JupyterHub proxy:400] Adding default route for Hub: / => http://127.0.0.1:8081
21:34:51.585 [ConfigProxy] info: Adding route / -> http://127.0.0.1:8081
21:34:51.586 [ConfigProxy] info: Route added / -> http://127.0.0.1:8081
21:34:51.588 [ConfigProxy] info: 201 POST /api/routes/
[I 2020-08-26 21:34:51.589 JupyterHub app:2697] JupyterHub is now running at http://:8000
[I 2020-08-26 21:34:56.313 JupyterHub log:181] 302 GET / -> /hub/ (@::ffff:172.17.0.1) 1.81ms
[I 2020-08-26 21:34:56.331 JupyterHub log:181] 302 GET /hub/ -> /hub/login (@::ffff:172.17.0.1) 1.68ms
[E 2020-08-26 21:34:56.346 JupyterHub web:1788] Uncaught exception GET /hub/login (::ffff:172.17.0.1)
    HTTPServerRequest(protocol='http', host='localhost:8000', method='GET', uri='/hub/login', version='HTTP/1.1', remote_ip='::ffff:172.17.0.1')
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/tornado/web.py", line 1703, in _execute
        result = await result
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/login.py", line 135, in get
        self.finish(self._render(username=username))
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/login.py", line 93, in _render
        return self.render_template(
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 1163, in render_template
        template = self.get_template(name)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 1157, in get_template
        return self.settings['jinja2_env'].get_template(name)
      File "/usr/local/lib/python3.8/dist-packages/jinja2/environment.py", line 883, in get_template
        return self._load_template(name, self.make_globals(globals))
      File "/usr/local/lib/python3.8/dist-packages/jinja2/environment.py", line 857, in _load_template
        template = self.loader.load(self, name, globals)
      File "/usr/local/lib/python3.8/dist-packages/jinja2/loaders.py", line 429, in load
        raise TemplateNotFound(name)
    jinja2.exceptions.TemplateNotFound: login.html

[E 2020-08-26 21:34:56.349 JupyterHub web:1220] Uncaught exception in write_error
    Traceback (most recent call last):
      File "/usr/local/lib/python3.8/dist-packages/tornado/web.py", line 1218, in send_error
        self.write_error(status_code, **kwargs)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 1247, in write_error
        html = self.render_template('error.html', **ns)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 1163, in render_template
        template = self.get_template(name)
      File "/usr/local/lib/python3.8/dist-packages/jupyterhub/handlers/base.py", line 1157, in get_template
        return self.settings['jinja2_env'].get_template(name)
      File "/usr/local/lib/python3.8/dist-packages/jinja2/environment.py", line 883, in get_template
        return self._load_template(name, self.make_globals(globals))
      File "/usr/local/lib/python3.8/dist-packages/jinja2/environment.py", line 857, in _load_template
        template = self.loader.load(self, name, globals)
      File "/usr/local/lib/python3.8/dist-packages/jinja2/loaders.py", line 429, in load
        raise TemplateNotFound(name)
    jinja2.exceptions.TemplateNotFound: error.html

[E 2020-08-26 21:34:56.351 JupyterHub log:173] {
      "X-Forwarded-Host": "localhost:8000",
      "X-Forwarded-Proto": "http",
      "X-Forwarded-Port": "8000",
      "X-Forwarded-For": "::ffff:172.17.0.1",
      "Upgrade-Insecure-Requests": "1",
      "Cookie": "_ga=[secret]; csrftoken=[secret]; _xsrf=[secret]; username-localhost-8888=[secret]; username-localhost-18888=[secret]; username-localhost-58888=[secret]; username-localhost-32768=[secret]; username-localhost-32769=[secret]; username-localhost-32770=[secret]; sessionid_test73=[secret]; grafana_sess=[secret]; sessionid_test80=[secret]; sessionid_test81=[secret]; JSESSIONID=[secret]; sessionid_prod80a=[secret]; sessionid_idr=[secret]; sessionid_prod80b=[secret]; jupyterhub-session-id=[secret]; sessionid_prod83=[secret]; sessionid_test84=[secret]; sessionid_prod84=[secret]; sessionid_mini-idr0071=[secret]; sessionid_mini-idr0092=[secret]; sessionid_pilot-idr0092=[secret]; sessionid_pilot-idr0071=[secret]; sessionid_prod86=[secret]; sessionid_pilot-idr0090=[secret]; username-localhost-8889=[secret]; sessionid=[secret]",
      "Connection": "close",
      "Dnt": "1",
      "Accept-Encoding": "gzip, deflate",
      "Accept-Language": "en-GB,en;q=0.7,en-US;q=0.3",
      "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8",
      "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.13; rv:79.0) Gecko/20100101 Firefox/79.0",
      "Host": "localhost:8000"
    }
[E 2020-08-26 21:34:56.351 JupyterHub log:181] 500 GET /hub/login (@::ffff:172.17.0.1) 6.70ms
```

**Comment by minrk** [comment]
That's weird, can you do `pip freeze` and `find / -name login.html` in the container?

When I build from this PR, it works without any errors

**Comment by manics** [comment]
Did you have some files from an earlier build?

I get the missing template errors if I start from a *clean* repo:

    git clean -dxf
    docker build -t jupyterhub:clean --no-cache .
    docker run -it --rm -p 8000:8000 jupyterhub:clean

<details>

```
root@e610f0228c04:/srv/jupyterhub# which pip
/usr/local/bin/pip

root@e610f0228c04:/srv/jupyterhub# pip freeze
alembic @ file:///tmp/wheelhouse/alembic-1.4.2-py2.py3-none-any.whl
async-generator @ file:///tmp/wheelhouse/async_generator-1.10-py3-none-any.whl
attrs @ file:///tmp/wheelhouse/attrs-20.1.0-py2.py3-none-any.whl
certifi @ file:///tmp/wheelhouse/certifi-2020.6.20-py2.py3-none-any.whl
certipy @ file:///tmp/wheelhouse/certipy-0.1.3-py3-none-any.whl
cffi @ file:///tmp/wheelhouse/cffi-1.14.2-cp38-cp38-manylinux1_x86_64.whl
chardet @ file:///tmp/wheelhouse/chardet-3.0.4-py2.py3-none-any.whl
cryptography @ file:///tmp/wheelhouse/cryptography-3.1-cp35-abi3-manylinux2010_x86_64.whl
decorator @ file:///tmp/wheelhouse/decorator-4.4.2-py2.py3-none-any.whl
entrypoints @ file:///tmp/wheelhouse/entrypoints-0.3-py2.py3-none-any.whl
idna @ file:///tmp/wheelhouse/idna-2.10-py2.py3-none-any.whl
ipython-genutils @ file:///tmp/wheelhouse/ipython_genutils-0.2.0-py2.py3-none-any.whl
Jinja2 @ file:///tmp/wheelhouse/Jinja2-2.11.2-py2.py3-none-any.whl
jsonschema @ file:///tmp/wheelhouse/jsonschema-3.2.0-py2.py3-none-any.whl
jupyter-telemetry @ file:///tmp/wheelhouse/jupyter_telemetry-0.1.0-py3-none-any.whl
jupyterhub @ file:///tmp/wheelhouse/jupyterhub-1.2.0.dev0-py3-none-any.whl
Mako @ file:///tmp/wheelhouse/Mako-1.1.3-py2.py3-none-any.whl
MarkupSafe @ file:///tmp/wheelhouse/MarkupSafe-1.1.1-cp38-cp38-manylinux1_x86_64.whl
oauthlib @ file:///tmp/wheelhouse/oauthlib-3.1.0-py2.py3-none-any.whl
pamela @ file:///tmp/wheelhouse/pamela-1.0.0-py2.py3-none-any.whl
prometheus-client @ file:///tmp/wheelhouse/prometheus_client-0.8.0-py2.py3-none-any.whl
pycparser @ file:///tmp/wheelhouse/pycparser-2.20-py2.py3-none-any.whl
pycurl==7.43.0.2
pyOpenSSL @ file:///tmp/wheelhouse/pyOpenSSL-19.1.0-py2.py3-none-any.whl
pyrsistent @ file:///tmp/wheelhouse/pyrsistent-0.16.0-cp38-cp38-linux_x86_64.whl
python-dateutil @ file:///tmp/wheelhouse/python_dateutil-2.8.1-py2.py3-none-any.whl
python-editor @ file:///tmp/wheelhouse/python_editor-1.0.4-py3-none-any.whl
python-json-logger @ file:///tmp/wheelhouse/python_json_logger-0.1.11-py2.py3-none-any.whl
requests @ file:///tmp/wheelhouse/requests-2.24.0-py2.py3-none-any.whl
ruamel.yaml @ file:///tmp/wheelhouse/ruamel.yaml-0.16.10-py2.py3-none-any.whl
ruamel.yaml.clib @ file:///tmp/wheelhouse/ruamel.yaml.clib-0.2.0-cp38-cp38-manylinux1_x86_64.whl
six @ file:///tmp/wheelhouse/six-1.15.0-py2.py3-none-any.whl
SQLAlchemy @ file:///tmp/wheelhouse/SQLAlchemy-1.3.19-cp38-cp38-manylinux2010_x86_64.whl
tornado @ file:///tmp/wheelhouse/tornado-6.0.4-cp38-cp38-linux_x86_64.whl
traitlets @ file:///tmp/wheelhouse/traitlets-4.3.3-py2.py3-none-any.whl
urllib3 @ file:///tmp/wheelhouse/urllib3-1.25.10-py2.py3-none-any.whl

root@e610f0228c04:/srv/jupyterhub# find / -name login.html
/usr/local/share/jupyterhub/templates/login.html
```

</details>


If I run `python setup.py sdist bdist_wheel` (local python 3.8) before `docker build` it works:

    git clean -dxf
    python setup.py sdist bdist_wheel
    docker build -t jupyterhub:afterbuild --no-cache .
    docker run -it --rm -p 8000:8000 jupyterhub:afterbuild

<details>

```
root@6ec0eb32296c:/srv/jupyterhub# which pip
/usr/local/bin/pip

root@6ec0eb32296c:/srv/jupyterhub# pip freeze
alembic @ file:///tmp/wheelhouse/alembic-1.4.2-py2.py3-none-any.whl
async-generator @ file:///tmp/wheelhouse/async_generator-1.10-py3-none-any.whl
attrs @ file:///tmp/wheelhouse/attrs-20.1.0-py2.py3-none-any.whl
certifi @ file:///tmp/wheelhouse/certifi-2020.6.20-py2.py3-none-any.whl
certipy @ file:///tmp/wheelhouse/certipy-0.1.3-py3-none-any.whl
cffi @ file:///tmp/wheelhouse/cffi-1.14.2-cp38-cp38-manylinux1_x86_64.whl
chardet @ file:///tmp/wheelhouse/chardet-3.0.4-py2.py3-none-any.whl
cryptography @ file:///tmp/wheelhouse/cryptography-3.1-cp35-abi3-manylinux2010_x86_64.whl
decorator @ file:///tmp/wheelhouse/decorator-4.4.2-py2.py3-none-any.whl
entrypoints @ file:///tmp/wheelhouse/entrypoints-0.3-py2.py3-none-any.whl
idna @ file:///tmp/wheelhouse/idna-2.10-py2.py3-none-any.whl
ipython-genutils @ file:///tmp/wheelhouse/ipython_genutils-0.2.0-py2.py3-none-any.whl
Jinja2 @ file:///tmp/wheelhouse/Jinja2-2.11.2-py2.py3-none-any.whl
jsonschema @ file:///tmp/wheelhouse/jsonschema-3.2.0-py2.py3-none-any.whl
jupyter-telemetry @ file:///tmp/wheelhouse/jupyter_telemetry-0.1.0-py3-none-any.whl
jupyterhub @ file:///tmp/wheelhouse/jupyterhub-1.2.0.dev0-py3-none-any.whl
Mako @ file:///tmp/wheelhouse/Mako-1.1.3-py2.py3-none-any.whl
MarkupSafe @ file:///tmp/wheelhouse/MarkupSafe-1.1.1-cp38-cp38-manylinux1_x86_64.whl
oauthlib @ file:///tmp/wheelhouse/oauthlib-3.1.0-py2.py3-none-any.whl
pamela @ file:///tmp/wheelhouse/pamela-1.0.0-py2.py3-none-any.whl
prometheus-client @ file:///tmp/wheelhouse/prometheus_client-0.8.0-py2.py3-none-any.whl
pycparser @ file:///tmp/wheelhouse/pycparser-2.20-py2.py3-none-any.whl
pycurl==7.43.0.2
pyOpenSSL @ file:///tmp/wheelhouse/pyOpenSSL-19.1.0-py2.py3-none-any.whl
pyrsistent @ file:///tmp/wheelhouse/pyrsistent-0.16.0-cp38-cp38-linux_x86_64.whl
python-dateutil @ file:///tmp/wheelhouse/python_dateutil-2.8.1-py2.py3-none-any.whl
python-editor @ file:///tmp/wheelhouse/python_editor-1.0.4-py3-none-any.whl
python-json-logger @ file:///tmp/wheelhouse/python_json_logger-0.1.11-py2.py3-none-any.whl
requests @ file:///tmp/wheelhouse/requests-2.24.0-py2.py3-none-any.whl
ruamel.yaml @ file:///tmp/wheelhouse/ruamel.yaml-0.16.10-py2.py3-none-any.whl
ruamel.yaml.clib @ file:///tmp/wheelhouse/ruamel.yaml.clib-0.2.0-cp38-cp38-manylinux1_x86_64.whl
six @ file:///tmp/wheelhouse/six-1.15.0-py2.py3-none-any.whl
SQLAlchemy @ file:///tmp/wheelhouse/SQLAlchemy-1.3.19-cp38-cp38-manylinux2010_x86_64.whl
tornado @ file:///tmp/wheelhouse/tornado-6.0.4-cp38-cp38-linux_x86_64.whl
traitlets @ file:///tmp/wheelhouse/traitlets-4.3.3-py2.py3-none-any.whl
urllib3 @ file:///tmp/wheelhouse/urllib3-1.25.10-py2.py3-none-any.whl

root@6ec0eb32296c:/srv/jupyterhub# find / -name login.html
/usr/local/share/jupyterhub/templates/login.html
```

</details>


If I switch back to `master` building from a *clean* repo works:

    git clean -dxf
    docker build -t jupyterhub:master --no-cache .
    docker run -it --rm -p 8000:8000 jupyterhub:master

<details>

```
root@cd0c6f59c22c:/srv/jupyterhub# which pip
/usr/local/bin/pip

root@cd0c6f59c22c:/srv/jupyterhub# pip freeze
alembic @ file:///tmp/wheelhouse/alembic-1.4.2-py2.py3-none-any.whl
async-generator @ file:///tmp/wheelhouse/async_generator-1.10-py3-none-any.whl
attrs @ file:///tmp/wheelhouse/attrs-20.1.0-py2.py3-none-any.whl
certifi @ file:///tmp/wheelhouse/certifi-2020.6.20-py2.py3-none-any.whl
certipy @ file:///tmp/wheelhouse/certipy-0.1.3-py3-none-any.whl
cffi @ file:///tmp/wheelhouse/cffi-1.14.2-cp36-cp36m-manylinux1_x86_64.whl
chardet @ file:///tmp/wheelhouse/chardet-3.0.4-py2.py3-none-any.whl
cryptography @ file:///tmp/wheelhouse/cryptography-3.1-cp35-abi3-manylinux2010_x86_64.whl
decorator @ file:///tmp/wheelhouse/decorator-4.4.2-py2.py3-none-any.whl
entrypoints @ file:///tmp/wheelhouse/entrypoints-0.3-py2.py3-none-any.whl
idna @ file:///tmp/wheelhouse/idna-2.10-py2.py3-none-any.whl
importlib-metadata @ file:///tmp/wheelhouse/importlib_metadata-1.7.0-py2.py3-none-any.whl
ipython-genutils @ file:///tmp/wheelhouse/ipython_genutils-0.2.0-py2.py3-none-any.whl
Jinja2 @ file:///tmp/wheelhouse/Jinja2-2.11.2-py2.py3-none-any.whl
jsonschema @ file:///tmp/wheelhouse/jsonschema-3.2.0-py2.py3-none-any.whl
jupyter-telemetry @ file:///tmp/wheelhouse/jupyter_telemetry-0.1.0-py3-none-any.whl
jupyterhub @ file:///tmp/wheelhouse/jupyterhub-1.2.0.dev0-py3-none-any.whl
Mako @ file:///tmp/wheelhouse/Mako-1.1.3-py2.py3-none-any.whl
MarkupSafe @ file:///tmp/wheelhouse/MarkupSafe-1.1.1-cp36-cp36m-manylinux1_x86_64.whl
oauthlib @ file:///tmp/wheelhouse/oauthlib-3.1.0-py2.py3-none-any.whl
pamela @ file:///tmp/wheelhouse/pamela-1.0.0-py2.py3-none-any.whl
prometheus-client @ file:///tmp/wheelhouse/prometheus_client-0.8.0-py2.py3-none-any.whl
pycparser @ file:///tmp/wheelhouse/pycparser-2.20-py2.py3-none-any.whl
pycurl==7.43.0.1
pyOpenSSL @ file:///tmp/wheelhouse/pyOpenSSL-19.1.0-py2.py3-none-any.whl
pyrsistent @ file:///tmp/wheelhouse/pyrsistent-0.16.0-cp36-cp36m-linux_x86_64.whl
python-dateutil @ file:///tmp/wheelhouse/python_dateutil-2.8.1-py2.py3-none-any.whl
python-editor @ file:///tmp/wheelhouse/python_editor-1.0.4-py3-none-any.whl
python-json-logger @ file:///tmp/wheelhouse/python_json_logger-0.1.11-py2.py3-none-any.whl
requests @ file:///tmp/wheelhouse/requests-2.24.0-py2.py3-none-any.whl
ruamel.yaml @ file:///tmp/wheelhouse/ruamel.yaml-0.16.10-py2.py3-none-any.whl
ruamel.yaml.clib @ file:///tmp/wheelhouse/ruamel.yaml.clib-0.2.0-cp36-cp36m-manylinux1_x86_64.whl
six @ file:///tmp/wheelhouse/six-1.15.0-py2.py3-none-any.whl
SQLAlchemy @ file:///tmp/wheelhouse/SQLAlchemy-1.3.19-cp36-cp36m-manylinux2010_x86_64.whl
tornado @ file:///tmp/wheelhouse/tornado-6.0.4-cp36-cp36m-linux_x86_64.whl
traitlets @ file:///tmp/wheelhouse/traitlets-4.3.3-py2.py3-none-any.whl
urllib3 @ file:///tmp/wheelhouse/urllib3-1.25.10-py2.py3-none-any.whl
zipp @ file:///tmp/wheelhouse/zipp-3.1.0-py3-none-any.whl

root@cd0c6f59c22c:/srv/jupyterhub# find / -name login.html
/usr/local/share/jupyterhub/templates/login.html
```

</details>

**Comment by minrk** [comment]
@manics, sorry for missing this. You're right, I don't know what changed, whether it's in Python3.8 or pip, but the result is that `pip wheel .` does not include generated files (components, css). This has been fiddly in the past. However, running `setup.py bdist_wheel` explicitly resolves it. I can't tell you what changed or why, exactly.

**Comment by minrk** [comment]
added a test to verify that the static files are present in the dockerfiles, so it should be all set. Thanks @manics!

**Comment by mariusvniekerk** [comment]
I have noticed this behavior when upgrading a very old jupyterhub to the current version where it was being caused by

`jupyterlab.labhubapp` being used instead of `jupyterhub.singleuser.app` in your singleuser launch script.

**Pull request merged**